File: src/Arqanore/Input/Joystick.cs
using System;
using System.Runtime.InteropServices;
using Arqan;

namespace Arqanore.Input
{
    public static class Joystick
    {
        public static bool Connected(int joystick)
        {
            return GLFW.glfwJoystickPresent(joystick) == 1;
        }

        public static float[] GetAxis(int joystick)
        {
            var arraySize = 0;
            var ptr = GLFW.glfwGetJoystickAxes(joystick, out arraySize);
            var states = new float[arraySize];
            
            Marshal.Copy(ptr, states, 0, arraySize);
            return states;
        }

        public static bool[] GetButtons(int joystick)
        {
            var arraySize = 0;
            var ptr = GLFW.glfwGetJoystickButtons(joystick, out arraySize);
            var states = new byte[arraySize];
            var result = new bool[states.Length];
            
            Marshal.Copy(ptr, states, 0, arraySize);

            for (var i=0; i<states.Length; i++)
            {
                if (states[i] == 1) result[i] = true;
            }

            return result;
        }

        public static byte[] GetHats(int joystick)
        {
            var arraySize = 0;
            var ptr = GLFW.glfwGetJoystickHats(joystick, out arraySize);
            var states = new byte[arraySize];
            
            Marshal.Copy(ptr, states, 0, arraySize);
            return states;
        }

        public static string GetName(int joystick)
        {
            return GLFW.glfwGetJoystickName(joystick);
        }
    }
}

File: src/arqanore/joystick.cpp
#include <stdexcept>
#include <GLFW/glfw3.h>
#include "arqanore/joystick.h"
#include "arqanore/exceptions.h"

bool arqanore::Joystick::is_connected(int jid) {
    if (jid < 0 || jid > 15) {
        throw ArqanoreException("Illegal joystick id. Expected a number between or equal to 0 and 15.");
    }

    return glfwJoystickPresent(jid);
}


std::string arqanore::Joystick::get_name(int jid) {
    auto input = glfwGetJoystickName(jid);

    if (input == nullptr) {
        throw ArqanoreException("Joystick " + std::to_string(jid) + " is not connected.");
    }

    return std::string(input);
}

std::string arqanore::Joystick::get_guid(int jid) {
    auto input = glfwGetJoystickGUID(jid);

    if (input == nullptr) {
        throw ArqanoreException("Joystick " + std::to_string(jid) + " is not connected.");
    }

    return std::string(input);
}

std::vector<float> arqanore::Joystick::get_axes(int jid) {
    int *arr_length = new int;
    auto arr = glfwGetJoystickAxes(jid, arr_length);
    std::vector<float> result;

    if (arr == nullptr) {
        throw ArqanoreException("Joystick " + std::to_string(jid) + " is not connected.");
    }

    for (int i = 0; i < *arr_length; i++) {
        result.push_back(arr[i]);
    }

    delete arr_length;
    return result;
}

std::vector<int> arqanore::Joystick::get_buttons(int jid) {
    int *arr_length = new int;
    auto arr = glfwGetJoystickButtons(jid, arr_length);
    std::vector<int> result;

    if (arr == nullptr) {
        throw ArqanoreException("Joystick " + std::to_string(jid) + " is not connected.");
    }

    for (int i = 0; i < *arr_length; i++) {
        result.push_back(arr[i]);
    }

    delete arr_length;
    return result;
}
